project_setup/scripts/iam-to-github

```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env python
+
+"""Create GitHub secrets from AWS credentials extracted from Terraform state.
+
+This command must be executed in the directory containing the .terraform state
+within a GitHub project.  It will attempt to detect the repository name from
+the projects git origin.
+
+It requires a Personal Access Token from GitHub that has "repo" access scope.
+
+Usage:
+  iam-to-github [options] <github-personal-access-token>
+  iam-to-github (-h | --help)
+
+Options:
+  -h --help              Show this message.
+  --log-level=LEVEL      If specified, then the log level will be set to
+                         the specified value.  Valid values are "debug", "info",
+                         "warning", "error", and "critical". [default: info]
+  --repo=REPONAME        Use provided repository name instead of detecting it.
+  --suffix=SUFFIX        Append a suffix to the secret names.
+  --user=FILTER          Only process users that contain the filter text.
+"""
+
+# Standard Python Libraries
+from base64 import b64encode
+import json
+import logging
+import re
+import subprocess  # nosec : security implications have been considered
+import sys
+from typing import Any, Dict, Generator, List, Optional, Tuple
+
+# Third-Party Libraries
+import docopt
+from nacl import encoding, public
+import requests
+from schema import And, Or, Schema, SchemaError, Use
+
+GIT_URL_RE: re.Pattern = re.compile("git@github.com:(.*).git")
+
+
+def creds_from_child(child_module: Dict) -> Generator[Tuple[str, str, str], None, None]:
+    """Search for IAM access keys in child resources.
+
+    Yields (user, key_id, secret) when found.
+    """
+    for resource in child_module["resources"]:
+        if resource["address"] == "aws_iam_access_key.key":
+            key_id: str = resource["values"]["id"]
+            secret: str = resource["values"]["secret"]
+            user: str = resource["values"]["user"]
+            yield user, key_id, secret
+    return
+
+
+def creds_from_terraform() -> Generator[Tuple[str, str, str], None, None]:
+    """Retrieve IAM credentials from terraform state.
+
+    Yields (user, key_id, secret) when found.
+    """
+    user: str
+    key_id: str
+    secret: str
+    c = subprocess.run(["terraform", "show", "--json"], stdout=subprocess.PIPE)  # nosec
+    # Normally we'd check the process return code here.  But terraform is perfectly
+    # happy to return zero even if there was no state files.
```

Comment:
Typo in comment: Should be "file", not "files".

Reply:
There are multiple files that make up the state.  I changed the sentence to the subjunctive mood.

Reply:
Fixed in: 381953d